pkg/sql/pg_catalog_test.go

```diff
@@ -0,0 +1,379 @@
+package sql
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/errors/oserror"
+)
+
+// This constants enumerates indexes in pg_catalog_tables
+const (
+	testTableIndex = iota
+	testColumnIndex
+	testTypeIndex
+	testTypeOidIndex
+)
+
+// This query uses udt_name::regtype instead of data_type column because
+// data_type only says "ARRAY" but does not say which kind of array it is.
+const getPgCatalogSQL = `
+	SELECT table_name, column_name, udt_name::regtype data_type, udt_name::regtype::Oid data_type_oid
+	FROM information_schema.columns 
+	WHERE table_schema = 'pg_catalog'
+	ORDER BY 1, 2;
+`
+
+// Test data files
+const (
+	pgCatalogDump = "pg_catalog_tables"          // PostgreSQL pg_catalog schema
+	expectedDiffs = "pg_catalog_test-diffs.json" // Contains expected difference between postgres and cockroach
+	testdata      = "testdata"                   // testdata directory
+)
+
+// When running test with -rewrite-diffs test will pass and re-create pg_catalog_test-diffs.json
+var rewriteFlag = flag.Bool("rewrite-diffs", false, "This will re-create the expected diffs file")
+
+// summary will keep accountability for any unexpected difference and report it in the log
+type summary struct {
+	missingTables        int
+	missingColumns       int
+	mismatchDatatypesOid int
+}
+
+// TestColumn is a structure which contains a small description about the datatype of a column, but this can also be
+// used as a diff information if populating ExpOid and ExpDataType. Fields are exported for Marshaling purposes.
+type TestColumn struct {
+	Oid         uint32  `json:"oid"`
+	DataType    string  `json:"dataType"`
+	ExpOid      *uint32 `json:"expOid"`
+	ExpDataType *string `json:"expDataType"`
+}
+
+// testColumns maps column names with datatype description
+type testColumns map[string]*TestColumn
+
+// testTable have 2 use cases:
+// First: This is used to model pg_schema for postgres and cockroach db for comparing purposes by mapping tableNames
+// with columns.
+// Second: This is used to store and load expected diffs:
+// - Using it this way, a table name pointing to a zero length testColumns means that we expect this table to be missing
+//   in cockroach db
+// - If testColumns is not empty but columnName points to null, we expect that column to be missing in that table in
+//   cockroach db
+// - If column Name points to a not null TestColumn, the test column describes how we expect that data type to be
+//   different between cockroach db and postgres
+type testTable map[string]testColumns
+
+// addRow is used to load data from postgres or cockroach pg_catalog schema
+func (pgTables testTable) addRow(
```

Comment:
i think it should be renamed to:

```
func (p *PGCatalogTable) addColumnMetadata(
  ...
) {
```

we have a convention of only using one or two letters to name the function receiver, and i think in this case it should be a pointer

Reply:
Done